mithril-common/src/chain_reader/pallas_chain_reader.rs

```diff
@@ -0,0 +1,262 @@
+    /// Creates a new `PallasChainReader` with the specified socket and network.
+    pub fn new(socket: &Path, network: CardanoNetwork) -> Self {
+        Self {
+            socket: socket.to_owned(),
+            network,
+            client: None,
+        }
+    }
+
+    /// Creates and returns a new `NodeClient` connected to the specified socket.
+    async fn new_client(&self) -> StdResult<NodeClient> {
+        let magic = self.network.code();
+        NodeClient::connect(&self.socket, magic)
+            .await
+            .map_err(|err| anyhow!(err))
+            .with_context(|| "PallasChainReader failed to create a new client")
+    }
+
+    async fn get_client(&mut self) -> StdResult<&mut NodeClient> {
+        if self.client.is_none() {
+            self.client = Some(self.new_client().await?);
+        }
+
+        Ok(self.client.as_mut().unwrap())
+    }
+}
+
+impl Drop for PallasChainReader {
+    fn drop(&mut self) {
+        if let Some(client) = self.client.take() {
+            tokio::spawn(async move {
+                let _ = client.abort().await;
+            });
+        }
+    }
+}
+
+#[async_trait]
+impl ChainBlockReader for PallasChainReader {
+    async fn process_next_chain_block(
+        &mut self,
+        next: chainsync::NextResponse<chainsync::BlockContent>,
+    ) -> StdResult<Option<ChainBlockNextAction>> {
+        match next {
+            NextResponse::RollForward(raw_block, forward_tip) => {
+                Ok(Some(ChainBlockNextAction::RollForward {
+                    next_point: forward_tip.into(),
+                    raw_block,
+                }))
+            }
+            NextResponse::RollBackward(rollback_point, _) => {
+                Ok(Some(ChainBlockNextAction::RollBackward {
+                    rollback_point: rollback_point.into(),
+                }))
+            }
+            NextResponse::Await => Ok(None),
+        }
+    }
+
+    async fn get_next_chain_block(
+        &mut self,
+        point: &ChainPoint,
+    ) -> StdResult<Option<ChainBlockNextAction>> {
+        let client = self.get_client().await?;
+        let chainsync = client.chainsync();
+
+        chainsync
+            .find_intersect(vec![point.to_owned().into()])
+            .await?;
+
+        let next = match chainsync.has_agency() {
+            true => chainsync.request_next().await?,
+            false => chainsync.recv_while_must_reply().await?,
+        };
+
+        self.process_next_chain_block(next).await
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::fs;
+
+    use pallas_network::{
+        facades::NodeServer,
+        miniprotocols::{
+            chainsync::{BlockContent, Tip},
+            Point,
+        },
+    };
+    use tokio::net::UnixListener;
+
+    use super::*;
+
+    use crate::test_utils::TempDir;
+
+    /// Returns a fake chain point for testing purposes.
+    fn get_fake_chain_point_backwards() -> ChainPoint {
+        ChainPoint::from(Point::Specific(
```

Comment:
Maybe the `PointSpecific` value which is repeated 3 times could be shared somewhere?